product_attribute_global_item_code/product.py

```diff
@@ -0,0 +1,84 @@
+# -*- coding: utf-8 -*-
+###############################################################################
+#
+#   Copyright (C) 2015 Akretion (http://www.akretion.com). All Rights Reserved
+#   @author Abdessamad HILALI <abdessamad.hilali@akretion.com>
+#
+#   This program is free software: you can redistribute it and/or modify
+#   it under the terms of the GNU Affero General Public License as
+#   published by the Free Software Foundation, either version 3 of the
+#   License, or (at your option) any later version.
+#
+#   This program is distributed in the hope that it will be useful,
+#   but WITHOUT ANY WARRANTY; without even the implied warranty of
+#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#   GNU Affero General Public License for more details.
+#
+#   You should have received a copy of the GNU Affero General Public License
+#   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+###############################################################################
+
+from openerp import fields, models, api, _
+from openerp.exceptions import Warning
+
+
+class ProductTemplate(models.Model):
+    _inherit = "product.template"
+
+    base_code = fields.Char('Base Code',
+                            help="this field is used as a base "
+                            "for automatically create "
+                            "Internal Reference(default_code)")
+
+
+class ProductProduct(models.Model):
+    _inherit = "product.product"
+
+    use_internal_ref = fields.Boolean('Manual Internal Reference',
+                                      default=False)
+    manual_default_code = fields.Char(
+        help="This is an invisible field used to store default_code value"
+    )
+    default_code = fields.Char(compute="_compute_default_code",
+                               inverse="_set_manual_default_code",
+                               store=True)
+
+    @api.multi
+    def _build_code(self, vals):
+        """"inherit this method to customize your own behavior"""
+        self.ensure_one()
+        return "".join([self.base_code] + ["".join(val) for val in vals])
+
+    @api.multi
+    def _get_default_code(self):
+        self.ensure_one()
+        vals = []
+        for value in self.attribute_value_ids:
+            if not value.attribute_id.code:
+                raise Warning(
+                    _("Missing code for %s to correct this error you must "
+                      "see :Configuration  -> Product Categories -> Attribute")
+                    % (value.attribute_id.name))
+            if not value.code:
+                raise Warning(
```

Comment:
This function field cause an issue with existing data.  You end up in a circle.  If you add a code to the attribute you cannot save because it wants a code on the values.  Likewise if you add on the values, it fails because you require a code on the attribute.